### PR TITLE
fw_pos_control: purge L1 controller

### PR DESCRIPTION
**Pull request by dagar**
Opening this to identify what's required before we can fully switch the FW position controller over to NPFG and remove L1.

#### Optional:
 - [ ] reset roll integrator for discrete switch in position setpoint, remove "circle mode" handling?
 - [ ] do we want a parameter to disable the use of wind estimates entirely? (even if they are available)?
 - [ ] how do we determine if wind estimates are "valid"?

### CODE REVIEW

File: ROMFS/px4fmu_common/init.d-posix/airframes/1030_plane [hunk 1]
```diff
@@ -11,7 +11,6 @@ param set-default EKF2_MAG_YAWLIM 0
 param set-default FW_LND_AIRSPD_SC 1
 param set-default FW_LND_ANG 8
 
-param set-default FW_L1_PERIOD 12
```
Comment (dagar) [inline ROMFS/px4fmu_common/init.d-posix/airframes/1030_plane:14]: TODO: param translation to NPFG_PERIOD

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 2]
```diff
@@ -536,67 +530,57 @@ FixedwingPositionControl::tecs_status_publish()
...
+	npfg_status.signed_track_error = _npfg.getTrackError();
+	npfg_status.track_error_bound = _npfg.getTrackErrorBound();
+	npfg_status.airspeed_ref = _npfg.getAirspeedRef();
+	npfg_status.min_ground_speed_ref = _npfg.getMinGroundSpeedRef();
+	npfg_status.adapted_period = _npfg.getAdaptedPeriod();
+	npfg_status.p_gain = _npfg.getPGain();
+	npfg_status.time_const = _npfg.getTimeConst();
+	npfg_status.timestamp = hrt_absolute_time();
+
+	_npfg_status_pub.publish(npfg_status);
```
Comment (Jaeyoung-Lim) [inline src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:583]: Do we actually need a separate `npfg_status`? or can we merge it in to pos control status?

Comment (tstastny) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:583]: maybe one or the other - if there were actually more position controllers that have common outputs.. the generic name makes sense. but for the moment.. maybe just directly `npfg_status` stays.. and we kill `pos_control_status`? it has informative fields for this specific controller.

File: ROMFS/px4fmu_common/init.d-posix/airframes/1030_plane (hunk 1 above)
Comment (KonradRudin) [reply ROMFS/px4fmu_common/init.d-posix/airframes/1030_plane]: @tstastny is there a relation between the NPFG period and the L1 period we could leverage to transform them

File: src/modules/fw_path_navigation/FixedwingPositionControl.cpp [hunk 3]
```diff
@@ -1068,7 +1036,7 @@ FixedwingPositionControl::control_auto_position(const float control_interval, co
 
 	} else {
 		// No valid previous waypoint, go for the current wp.
-		// This is automatically handled by the L1/NPFG libraries.
+		// This is automatically handled by the NPFG libraries.
```
Comment (tstastny) [inline src/modules/fw_path_navigation/FixedwingPositionControl.cpp:1039]: I would remove the hacky superpositioned waypoints here and rather either call navigateWaypoints() (when there is a valid previous point) or navigateBearing() when there is only the one point (calculating the bearing from aircraft current position to singular waypoint).

File: src/modules/fw_path_navigation/FixedwingPositionControl.cpp [hunk 4]
```diff
@@ -1245,7 +1201,7 @@ FixedwingPositionControl::control_auto_loiter(const float control_interval, cons
...
```
Comment (tstastny) [inline src/modules/fw_path_navigation/FixedwingPositionControl.cpp:1204]: previous waypoint is not used in this method. would delete this entire if condition.

File: src/modules/fw_path_navigation/FixedwingPositionControl.cpp [hunk 5]
```diff
@@ -2364,35 +2235,6 @@ FixedwingPositionControl::reset_landing_state()
...
 }
 
-bool FixedwingPositionControl::using_npfg_with_wind_estimate() const
-{
-	// high wind mitigation is handled by NPFG if the wind estimate is valid
-	return _param_fw_use_npfg.get() && _wind_valid;
-}
-
-Vector2f
-FixedwingPositionControl::get_nav_speed_2d(const Vector2f &ground_speed)
```
Comment (tstastny) [inline src/modules/fw_path_navigation/FixedwingPositionControl.cpp:2374]: why do you delete this method? i dont particularly like the air/ground angle logic, is a bit clunky, but it is the last defense in high wind when we dont have wind estimate. of course, one could also argue that if we have a valid airspeed estimate.. we should probably still have valid wind (or maybe not if GNSS is out?)

**Comment by tstastny** [comment]
> @tstastny can you confirm, that the last commit here is legit?
https://github.com/PX4/PX4-Autopilot/commit/02a06df044a0fbe312edf5441e6c3f1aa67f2244

made a few comments on this commit above.

File: src/modules/fw_path_navigation/FixedwingPositionControl.cpp (hunk 5 above)
Comment (KonradRudin) [reply src/modules/fw_path_navigation/FixedwingPositionControl.cpp:2374]: That,s what i wanted to discuss with you, but i thought this was only done because L1 would get unstable, when we move "backwards" with respect to ground. In this case, it replaces the ground speed with the airspeed, but this gets then fed into the npfg navigate functions as ground speed again?!? Doesn't that screw up the npfg calculations? Also the npfg does internally account for that in the course to heading conversion or not? Also, this function is not always used, but for me only at random places.

File: src/modules/fw_path_navigation/FixedwingPositionControl.cpp (hunk 3 above)
Comment (KonradRudin) [reply src/modules/fw_path_navigation/FixedwingPositionControl.cpp:1039]: I fix it, but instead of navigateBearing i do navigateWaypoints where the previous waypoint is the aircraft position. Should be very similar to the navigateBearing function if the dt is small and would at least make sure to hit the current waypoint. Similar to the changes made in https://github.com/PX4/PX4-Autopilot/pull/21071

File: src/modules/fw_path_navigation/FixedwingPositionControl.cpp (hunk 4 above)
Comment (KonradRudin) [reply src/modules/fw_path_navigation/FixedwingPositionControl.cpp:1204]: Hmm, this made me think, the navigateLoiter function is not using the previous waypoint. But don't we actually want to fly on the line from the previous waypoint to the center circle, as long as we are "far" away from the circle?

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp (hunk 2 above)
Comment (KonradRudin) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:539]: Hey, @tstastny ans @Jaeyoung-Lim. If you don't mind, i would leave this as is for the moment. When we split up the lateral guidance and add the course, acceleration, etc. interface, we can consolidate them there and split them up as appropriate for the sub components.

File: src/modules/fw_path_navigation/FixedwingPositionControl.cpp (hunk 5 above)
Comment (tstastny) [reply src/modules/fw_path_navigation/FixedwingPositionControl.cpp:2374]: >  but i thought this was only done because L1 would get unstable, when we move "backwards" with respect to ground

yes, but we're considering a case here where there is no wind estimate, so the ground speed is the only thing getting input to NPFG. So effectively NPFG is no better than L1.. because it doesnt think there is any wind.

> but this gets then fed into the npfg navigate functions as ground speed again?!?

yeah it's a hack. but results in something similar to what NPFG does for backwards flight.

Now that I re-think about it though .. if we have an airspeed measurement, and a ground velocity estimate.. there is ZERO reason we shouldn't have at least a cursory wind estimate.

I'm good with removing the `nav_speed_2d()` method then. But then we need to check what we do with no wind estimate. What I'd propose...
- clarify if there is ever a situation where we have airspeed yaw and GNSS measurements, but no wind estimate @sfuhrer @bresch - if there is .. we can consider how to handle this (e.g. if we combine the measurements ourselves in the fw pos ctrl.. but that seems silly)
- in the case of no valid wind estimate (this ideally is only ever true when we are in AIRSPEEDLESS mode), feed in ground speed directly with zero for the wind vector. But then also adapt the minimum ground speed settings in NPFG (if they are not already set to some non-zero value). Then NPFG will do its best to maintain some non-zero ground speed despite not having the yaw/airspeed vector.
- beyond this, we cannot do any better, but at least we don't need the additional nav_speed_2d method.

File: src/modules/fw_path_navigation/FixedwingPositionControl.cpp (hunk 4 above)
Comment (tstastny) [reply src/modules/fw_path_navigation/FixedwingPositionControl.cpp]: > But don't we actually want to fly on the line from the previous waypoint to the center circle, as long as we are "far" away from the circle?

I'd argue not. I don't see any benefit to it. Rather just take the shortest path to the orbit point. But if we want this behavior, I think we should encode it outside NPFG methods. So if we want this as optional.. i dont know if there is param for this or not.. the navigate method could take the line between the prev and current waypoint until reaching `switchDistance()` + `radius` from the orbit center.

Comment (KonradRudin) [reply src/modules/fw_path_navigation/FixedwingPositionControl.cpp]: For sure, this should be outside the npfg. For the moment it would reside inside the auto loiter function. I also did a small test on main, here is the screenshot. 
Screenshot from 2023-02-07 16-55-03_small
So the last loiter is flown to directly, i would have thought it follows the yellow line (in the plan view, it also says it follows the blue line). I would say, if you want to use the shortest path, make the previous invalid, else follow the line, as it is more intuitive as what is shown as mission. What do you guys think?

Comment (KonradRudin) [reply src/modules/fw_path_navigation/FixedwingPositionControl.cpp]: Anyhow, this would be a separate pr and not in here.

Comment (sfuhrer) [reply src/modules/fw_path_navigation/FixedwingPositionControl.cpp]: I think we can get more consistency by always flying directly from the current position to the loiter, instead of from the previous waypoint. Waypoint line following for me only seems important for normal waypoints (eg mapping). @Jaeyoung-Lim what's your take on it?

File: src/modules/fw_path_navigation/FixedwingPositionControl.cpp (hunk 5 above)
Comment (sfuhrer) [reply src/modules/fw_path_navigation/FixedwingPositionControl.cpp:2374]: >clarify if there is ever a situation where we have airspeed yaw and GNSS measurements, but no wind estimate @sfuhrer @bresch - if there is .. we can consider how to handle this (e.g. if we combine the measurements ourselves in the fw pos ctrl.. but that seems silly)

Currently the EKF wind estimate can be switched off, and it's hard to prevent that unless we eg have beta fusion forcefully enabled always, and EKF2_ARSP_THR set below FW_AIRSPD_TRIM or so.
If we want to be independent of the EKF wind estimates then we could add another standalone wind estimator within the FW guidance module, which we then could also independently tune and keep enabled always. 
I would rather rely on the EKF wind estimate, tune it to be more dynamic and at least give a warning if EKF2_ARSP_THR set below FW_AIRSPD_TRIM and EKF2_FUSE_BETA is disabled.

File: ROMFS/px4fmu_common/init.d-posix/airframes/1030_plane (hunk 1 above)
Comment (sfuhrer) [reply ROMFS/px4fmu_common/init.d-posix/airframes/1030_plane]: I thought they are comparable, so what about a 1:1 param translation?

Also FW_L1_R_SLEW_MAX should be translated to FW_PN_R_SLEW_MAX.

File: src/modules/fw_path_navigation/FixedwingPositionControl.cpp (hunk 5 above)
Comment (tstastny) [reply src/modules/fw_path_navigation/FixedwingPositionControl.cpp:2374]: I'd prefer not to re-invent the wheel with state estimation inside the controller.. since it should already be feasible in the EKF. Also.. don't like the idea of using the measurements directly (compatibility / consistency for feedback states to the controller).

So for the case that EKF wind estimate is (for whatever reason) turned off - I suppose this would mean assuming zero wind and bumping the min ground speed to non-zero value. Best then we could do is advise.. don't turn off the wind estimate..?

Otherwise we are just needlessly re-implementing a worse wind estimation in the controller if we use the measurements directly.

wdyt @sfuhrer @KonradRudin ?

File: src/modules/fw_path_navigation/FixedwingPositionControl.cpp (hunk 4 above)
Comment (tstastny) [reply src/modules/fw_path_navigation/FixedwingPositionControl.cpp]: > Waypoint line following for me only seems important for normal waypoints (eg mapping).

I'd also parrot this sentiment. That if a specific line is wished to be flown.. actually plan that specific line with two waypoints. The root issue here is our improper display on QGC, and generally ambiguous behavioral intent. The line from last waypoint in mission to orbit point is misleading. We should rather fix that - and be explicit about what we are intending.

File: ROMFS/px4fmu_common/init.d-posix/airframes/1030_plane (hunk 1 above)
Comment (tstastny) [reply ROMFS/px4fmu_common/init.d-posix/airframes/1030_plane]: @KonradRudin yes - they are identical.

File: src/modules/fw_path_navigation/FixedwingPositionControl.cpp (hunk 5 above)
Comment (sfuhrer) [reply src/modules/fw_path_navigation/FixedwingPositionControl.cpp:2374]: I don't see a reason why a wind estimator in here would have worse performance, but I don't like the complexity, load and weight it adds. That's why my recommendation is the one noted above 
"I would rather rely on the EKF wind estimate, tune it to be more dynamic and at least give a warning if EKF2_ARSP_THR set below FW_AIRSPD_TRIM and EKF2_FUSE_BETA is disabled."
The param check and warning could reside in the FW guidance module. 
EKF2_WIND_NSD then should be increased, and the wind estimate instead be low passed filtered locally where required (eg for the "wind speed high" warning/RTL).

File: src/modules/fw_path_navigation/FixedwingPositionControl.cpp (hunk 4 above)
Comment (KonradRudin) [reply src/modules/fw_path_navigation/FixedwingPositionControl.cpp]: Fine by me, even better if the line in GGC vanished, then its also clearer from the mission planning side.

File: src/modules/fw_path_navigation/FixedwingPositionControl.cpp (hunk 5 above)
Comment (tstastny) [reply src/modules/fw_path_navigation/FixedwingPositionControl.cpp:2374]: > EKF2_WIND_NSD then should be increased, and the wind estimate instead be low passed filtered locally where required (eg for the "wind speed high" warning/RTL).

yes pls.

> give a warning if EKF2_ARSP_THR set below FW_AIRSPD_TRIM and EKF2_FUSE_BETA is disabled

that sounds fine. - but just to be clear, in this case that means the wind estimate is zero? or just possibly not a great estimate?

> I don't see a reason why a wind estimator in here would have worse performance

It doesn't *have to have worse performance. it's just not necessary to implemented it again in the fixed-wing controller. (especially if doing what is there now.. where it literally just takes the filtered airspeed value and yaw measurement to do a cursory geometry calculation .. so no filtering for the wind estimate)

> So for the case that EKF wind estimate is (for whatever reason) turned off

This is the only open question. What we do when there is no wind estimate. Again, my contention would be.. do nothing except perhaps (if enabled) bump the min ground speed. But I'd like wind use in the controller be intentional on config. So explicitly saying "yes I want wind estimates on" "yes i want to use it in the controller" "yes i understand that if i turn it off.. the controller may do weird turn around things in the very low ground speed flight regime if I dont enable some minimum ground speed"

Comment (tstastny) [reply src/modules/fw_path_navigation/FixedwingPositionControl.cpp:2374]: > So instead of removing the function above we should make sure the ground speed is always positive?

I rethought about this - would actually not even do that. We just need to make it clear in documentation what the controller will do with what configuration. So if the user purposely does not set a positive minimum ground speed, and purposely turns off the wind estimate, I would say this is on them. 

TL;DR - would remove the method, and if wind is invalid, set to zero.

Comment (sfuhrer) [reply src/modules/fw_path_navigation/FixedwingPositionControl.cpp:2374]: So did we settle on removing the method here, and propose two things in a second PR:
- increase EKF wind process noise
- add check in FW controller to have EKF wind estimation enabled

Comment (tstastny) [reply src/modules/fw_path_navigation/FixedwingPositionControl.cpp:2374]: ^strong yes from my side.

**Review by sfuhrer — APPROVED** [review]

**Comment by KonradRudin** [comment]
@tstastny and @dagar from your side also good for merge?
